src/modules/recorder.module.ts

```diff
@@ -0,0 +1,373 @@
+
+    // Play joining audio so Discord starts sending us data
+    connection.play("./yo.wav")
+    
+    // Register listener
+    connection.on("speaking", (user, speaking) => {
+        onUserChangeState(channel.id, user, speaking)
+    })
+
+    let lastCleared: number = timeToOffset(buffer, CLEAR_INTERVAL_SEC)
+    // Start interval to clear buffer
+    const interval: NodeJS.Timeout = setInterval(() => {
+        const now: number = timeToOffset(buffer, CLEAR_INTERVAL_SEC)
+        clearBuffer(buffer, lastCleared, now)
+        lastCleared = now
+    }, CLEAR_INTERVAL_SEC * MILLIS)
+
+    voiceChannels.set(channel.id, {
+        connection: connection,
+        buffer: buffer,
+        interval: interval,
+        lastClipTime: new Date(0)
+    })
+}
+
+// Leave a voice channel and clean up connections / intervals
+async function leaveChannel(channel: VoiceChannel) {
+    // always fetch the channel via API as using data from the cache
+    // may lead to the bot being stuck in an empty channel
+    const fetchedChannel: VoiceChannel = await channel.fetch(true) as VoiceChannel
+    const voiceChannelData = voiceChannels.get(channel.id)
+    if (voiceChannelData && fetchedChannel.members.size == 1) {
+        const connection: VoiceConnection = voiceChannelData.connection
+        voiceChannelData.connection.disconnect()
+        clearInterval(voiceChannelData.interval)
+        voiceChannels.delete(channel.id)
+    }
+}
+
+function randomClipId() {
+    let id: string = ""
+    for (let i = 0; i < CLIP_ID_LEN; i++) {
+        id += CLIP_ID_CHARS[Math.floor(Math.random() * CLIP_ID_CHARS.length)]
+    }
+    return id
+}
+
+async function saveClip(channel: VoiceChannel, clipName: string, textChannel: TextChannel) {
+    const voiceChannelData = voiceChannels.get(channel.id)
+    if (!voiceChannelData) {
+        return
+    }
+
+    const now = new Date()
+    if (voiceChannelData.lastClipTime.getTime() + CLIP_COOLDOWN_SEC * MILLIS > now.getTime()) {
+        return await textChannel.send("Please wait a few seconds before clipping again.")
+    }
+
+    // Prepare buffer for upload
+    const buffer: Buffer = prepareBuffer(voiceChannelData.buffer, BUFFER_WRITE_SEC)
+    const clipId: string = randomClipId()
+    const name = clipName ? clipName : clipId
+    const params = {
+        Bucket: BUCKET,
+        Key: `${clipId}.${FILE_EXT}`,
+        Body: buffer
+    }
+
+    // Upload file to S3
+    s3.upload(params, (err: Error, data: Record<string, any>) => {
+        if (err) {
+            console.log(err)
+            return
+        }
+
+        // Write entry to database
+        const clip = {
+            id: clipId,
+            time: new Date(),
+            name: name,
+            url: data.Location,
+            clipstart: 0,
+            clipend: buffer.length,
+            participants: channel.members.map(member => member.id)
+        }
+
+        createClip(clip).then(async res => {

```

Comment:
Because this is all in series, I'd rather it stick to one style. Either use `.then` and `.catch` or `async/await` and `try/catch`.